File: lib/tiers.ts
import type { Tier } from "@/components/design-system/tier-badge";

const TIERS: readonly Tier[] = ["bronze", "silver", "gold", "platinum"];

/**
 * Lifetime-XP thresholds for each league tier (cumulative — a tier is reached
 * at this XP and never lost). Mirrors league_tier_for_xp in the DB trigger;
 * FORMULAS.md §3 is the source of truth. Keep all three in sync.
 */
export const LEAGUE_TIER_MIN_XP: Record<Tier, number> = {
  bronze: 0,
  silver: 5000,
  gold: 20000,
  platinum: 50000,
};

/** League tier from lifetime XP (cumulative thresholds). */
export function leagueTierFromXp(xp: number): Tier {
  const safe = Number.isFinite(xp) && xp > 0 ? xp : 0;
  if (safe >= LEAGUE_TIER_MIN_XP.platinum) return "platinum";
  if (safe >= LEAGUE_TIER_MIN_XP.gold) return "gold";
  if (safe >= LEAGUE_TIER_MIN_XP.silver) return "silver";
  return "bronze";
}

/**
 * Map the stored `profiles.league_tier` string onto the DS `TierBadge`
 * `Tier` union. Unknown or missing values fall back to `"bronze"`.
 */
export function tierFromLeague(leagueTier?: string | null): Tier {
  const normalized = leagueTier?.toLowerCase().trim();
  return TIERS.find((tier) => tier === normalized) ?? "bronze";
}

File: apps/api/internal/domain/tiers/tiers.go
// Package tiers maps lifetime XP to league tiers (FORMULAS.md §3).
package tiers

import "strings"

// Tier is a league tier, lowest to highest: bronze, silver, gold, platinum.
type Tier string

// The four tiers, lowest to highest.
const (
	Bronze   Tier = "bronze"
	Silver   Tier = "silver"
	Gold     Tier = "gold"
	Platinum Tier = "platinum"
)

// MinXP is each tier's cumulative lifetime-XP threshold; a tier is never lost.
var MinXP = map[Tier]int64{
	Bronze:   0,
	Silver:   5_000,
	Gold:     20_000,
	Platinum: 50_000,
}

// highestFirst is the order FromXP checks thresholds in.
var highestFirst = []Tier{Platinum, Gold, Silver, Bronze}

// FromXP returns the tier reached with this lifetime XP.
func FromXP(xp int64) Tier {
	for _, tier := range highestFirst {
		if xp >= MinXP[tier] {
			return tier
		}
	}
	return Bronze
}

// FromLeague normalizes a stored profiles.league_tier value; unknown or
// missing values are bronze.
func FromLeague(league *string) Tier {
	if league == nil {
		return Bronze
	}
	tier := Tier(strings.ToLower(strings.TrimSpace(*league)))
	if _, known := MinXP[tier]; known {
		return tier
	}
	return Bronze
}
